### PR TITLE
document the test infrastructure

### PR DESCRIPTION
**Pull request by nikomatsakis**
cc #10

This content is *mostly* taken from the `COMPILER_TESTS.md` in the compiler source tree, but I added some more notes.

### CODE REVIEW

File: src/tests/adding.md [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+**In general, we expect every PR that fixes a bug in rustc to come
+accompanied with a regression test of some kind.** This test should
+fail in master but pass after the PR. These tests are really useful
+for preventing us from repeating the mistakes of the past.
+
+To add a new test, the first thing you generally do is to create a
+file, typically a Rust source file. Test files have a particular
+structure:
+
+- They always begin with the copyright notice;
```
Comment (mgattozzi) [inline src/tests/adding.md]: Maybe it would be useful to link the copyright notice here!

Comment (nikomatsakis) [reply src/tests/adding.md]: ugh. I just wish it didn't exist. But I suppose you are right.

File: src/tests/adding.md [hunk 2]
```diff
@@ -0,0 +1,273 @@
+# Adding new tests
+
+**In general, we expect every PR that fixes a bug in rustc to come
+accompanied with a regression test of some kind.** This test should
```
Comment (mre) [inline src/tests/adding.md]: Super nitpicky, but "accompanied by" is more common. 🤓 
https://ell.stackexchange.com/a/109564

File: src/tests/running.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+on debuginfo, you may be better off with the debuginfo test suite:
+
+```bash
+./x.py test --stage 1 src/test/debuginfo
+```
+
+**Warning:** Note that bors only runs the tests with the full stage 2
+build; therefore, while the tests **usually** work fine with stage 1,
+there are some limitations. In particular, `./x.py test --stage 1`
+(ie.,
```
Comment (ehuss) [inline src/tests/running.md]: This statement seems incomplete.

File: src/tests/intro.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+called a **test suite** -- they house a group of tests that are run in
+a distinct mode.
+
+[`src/test`]: https://github.com/rust-lang/rust/tree/master/src/test
+
+Here is a brief summary of the test suites as of this writing and what
+they mean. In some cases, the test suites are linked to parts of the manual
+that give more details.
+
+- [`ui`](ui.html) -- tests that check the exact stdout/stderr from compilation
```
Comment (ehuss) [inline src/tests/intro.md]: I'm confused, what does this link to?

File: src/tests/running.md (hunk 3 above)
Comment (mark-i-m) [inline src/tests/running.md]: ??

**Comment by ehuss** [comment]
Thanks @nikomatsakis, I find this very useful!

There also seems to be some additional information captured at https://brson.github.io/2017/07/10/how-rust-is-tested that may be useful.  I was wondering if some of that can maybe be brought over?  I can help if desired, but my perspective is from someone who knows very little about how things work (other than reading this), so I'm uncertain if I'm qualified.

**Comment by mark-i-m** [comment]
@ehuss I also really like that blog post. I think the main concern is how to not plagiarize it... One idea would be to just mention it as a further reference.

>  I can help if desired, but my perspective is from someone who knows very little about how things work (other than reading this), so I'm uncertain if I'm qualified.

Don't worry too much about this. I also don't know that much (except about macro parsing :P). Any help is appreciated!

**Comment by mark-i-m** [comment]
@nikomatsakis I thought of a couple of other things that we could mention here:
- crater: what is it and why is it useful? What kinds of changes need a crater run?
- tidy

**Comment by QuietMisdreavus** [comment]
Whenever this lands, i'd like to add a chapter talking about the things specific to the rustdoc tests. If you'd like, you can crib some initial content from [the header of `htmldocck.py`](https://github.com/rust-lang/rust/blob/master/src/etc/htmldocck.py). That's the best documentation for rustdoc tests at this point, and i'd like to bring it out of the test script.

File: src/tests/adding.md (hunk 1 above)
Comment (mgattozzi) [reply src/tests/adding.md]: @nikomatsakis I totally understand your pain

**Comment by mark-i-m** [comment]
@QuietMisdreavus That's sounds great! Is there a reason why it must be done after this PR merges?

**Comment by QuietMisdreavus** [comment]
@mark-i-m Because i wanted to be the one to write it. :stuck_out_tongue_winking_eye: I suppose i could make it a PR-to-the-PR, but i don't want to hold up this PR with extra content.

**Comment by mark-i-m** [comment]
@QuietMisdreavus Got it :D

**Comment by ehuss** [comment]
Here are some other test-related things that might be useful to have.  Perhaps they could be added as separate PRs?  This guide seems to be very focused on the core of the compiler, so I'm uncertain how much you want it to talk about testing the standard library, and the tools and components that surround the compiler.

* More discussion of the testing infrastructure.  In particular, bors/homu, travis/appveyor, platform emulation, etc.  
* At least briefly mention the other tests that get run, such as:
    * Unittests.  In particular, how to run individual tests, roughly which libraries are automatically tested, etc.  So far the guide seems to focus on just the tests in the `test/` directory.
    * Doctests.  How to run the doc tests in the standard library.  Other doc testing steps like linkchecker.
    * distcheck
    * tidy
    * Mention that tests also get run for individual packages, such as bootstrap, cargo, rls, rustfmt, miri, clippy, cargotest.
* crater/cargobomb?
* Is fuzzing in active use by the team?
* How does performance testing enter the picture?

File: src/tests/running.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+```
+./x.py test --stage 1 src/test/ui --test-args issue-1234
+```
+
+Under the hood, the test runner invokes the standard rust test runner
+(the same one you get with `#[test]`), so this command would wind up
+filtering for tests that include "issue-1234" in the name.
+
+Often, though, it's easier to just run the test by hand. More tests are
```
Comment (phansch) [inline src/tests/running.md]: I guess this should be 'Most tests'?

**Comment by mark-i-m** [comment]
@nikomatsakis Ping :)

**Comment by nikomatsakis** [comment]
Hi =) Yes, I meant to do this last week, but stuff got away from me. Gonna take a crack at addressing the feedback right now.

File: src/tests/intro.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/tests/intro.md]: link was wrong, fixed, not sure about the CI question tho

**Comment by nikomatsakis** [comment]
Things I did not do (I will add them to the main issue, they seem like good follow-up):

- crater: what is it and why is it useful? What kinds of changes need a crater run?
  - maybe talk about "perf runs" too
- rustdoc tests 
- [these things](https://github.com/rust-lang-nursery/rustc-guide/pull/47#issuecomment-363968994)

**Comment by nikomatsakis** [comment]
I added those to the issue.

File: src/conventions.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+```
+
+<a name=cio>
+
+# Using crates from crates.io
+
+It is allowed to use crates from crates.io, though external
+dependencies should not be added gratuitously. All such crates must
+have a suitably permissive license. There is an automatic check which
+inspects the Cargo metadata to ensure this.
```
Comment (mark-i-m) [inline src/conventions.md]: Is there some sort of vetting that is done? Obviously, we don't want rustc's dependencies to explode and we don't want low-quality crates (or even malicious code).

Comment (nikomatsakis) [reply src/conventions.md]: There is not. There maybe should be.

Comment (nikomatsakis) [reply src/conventions.md]: See also https://internals.rust-lang.org/t/rustc-dependency-policy/6537

**Pull request merged**